docs/dg/dev/architecture/coding-conventions.md

```diff
@@ -0,0 +1,1385 @@
+   */
+  protected $productFacade;
+
+  public function __construct($productFacade)
+  {
+    $this->productFacade = $productFacade;
+  }
+
+  public function addProduct(ProductAbstractTransfer $productAbstractTransfer, array $productConcreteCollection): int
+  {
+    return $this->productFacade->addProduct($productAbstractTransfer, $productConcreteCollection);
+  }
+}
+```
+
+## Plugin
+
+**Description**
+- Plugins are classes which are used to realize Inversion-Of-Control - instead of a direct call to another module's facade, a `plugin` can be provided as an optional and configurable class.
+- Plugins are the only classes that can be instantiated across modules.
+- See more details about the implementation in [Plugin Interface](#plugin-interface) component.
+
+**Conventions**
+- The plugin MUST NOT contain business logic but delegates to the underlying facade.
+- Plugin method names MUST use words `pre` and `post` instead of `before`, `after`.
+- Plugin class names MUST use words `pre`, `post`, `create`, `update`, `delete`, instead of `creator`, `updater`, `deleter`.
+- The plugin MUST implement a [Plugin Interface](#plugin-interface) which is provided by an extension module.
+
+**Guidelines**
+- Plugin name should be unique and give and overview about the behaviour that the plugin delivers (consider developers searching a plugin in a catalog across the framework only by plugin name).
+
+**Example**
+
+```php
+  /**
+    * @see /F/Q/N/FooDependencyProvider::getFooPlugins()
+    * @see /F/Q/N/BarDependencyProvider::getBarPlugins()
+    */
+  class ZipZapPlugin extends AbstractPlugin implements FooPluginInterface, BarPluginInterface {}
+```
+
+## Plugin Interface
+
+**Description**
+
+- Modules which consume plugins need to define their requirements with an interface.
+- The plugin interface needs to be placed in an “extension module”.
+
+There are three modules involved:
+
+1. **Plugin definer** (aka extension module): The module that defines and holds the plugin interface (example: CompanyPostCreatePluginInterface in CompanyExtension module.
+2. **Plugin executor**: The module that demands the plugin(s) in its Dependency Provider (example: CompanyDependencyProvider::getCompanyPostCreatePlugins() in Company module)
+3. **Plugin providers**: The modules that implement a plugin (example: CompanyBusinessUnitCreatePlugin in CompanyBusinessUnit module)
+
+
+
+**Conventions**
+- Plugin interfaces MUST be defined in extension module.
+  - Extension modules MUST be suffixed with Extension and follow regular module architecture.
+  - Extension modules MUST NOT contain anything else but plugin interfaces and Shared layer.
+- Plugin interface MUST have a class specification with plugin-stack details.
+- Each method of a plugin receives a collection as an input argument.
+
+**Guidelines**
+
+- Operations on single items in plugin stack is forbidden except for the following reasons (check this RFC for more details):
```

Comment:
Don't forget to add the link here.

Reply:
I've removed instead. RFC should have been integrated already.